### ISSUE TITLE
About the lambdas and some fitting problems

### ISSUE THREAD

**Issue by kaneyxx** [KindXiaoming/pykan#264]
Hi again, I keep struggling on fitting some formulas. I'm curious about those lambdas in `model.train()`. Only I can barely recognize are `lamb` ($\lambda$ in eq. 2.20 in paper), `lamb_l1` ($\mu_1$ in eq. 2.20 in paper), `lamb_entropy` ($\mu_2$ in eq. 2.20 in paper). To my knowledge, if we increase `lamb_l1`, we can make network become more sparse. I don't know how to set them for that we can obtain better fitting results. Moreover, how about `lamb_coef` and `lamb_coefdiff`? `lamb_coef` I think it's a penalty strength for spline coefficient (scale_sp in KANLayer), but I'm not quite sure how to set `lamb_coefdiff`.

Could you give me some advices or share some experiences about this?

Another interesting finding:
I'm fitting a formula that I already simplified it to $y = (-1.154) * x_1 + (-0.203) * x_2$. I used hellokan example code to try fitting and it worked well. When I changed to real-world data (I have checked the data, it shows the error less than 1e-17 by using formula above), however, I cannot fit it well even though I use the same model and train settings. So, I'm curious about how this happened. Any thoughts? (oh, the differences between toy set and real set is the data distribution, real ones are skewed)

**Comment by KindXiaoming** [comment]
Hi, (define weight = 1D function)
`lamb_l1` penalizes weight (deal with each weight independently), `lamb_entropy` reduces entropy of weights (a proxy for the number of active weights, considering weight interaction), `lamb_coef` encourages weight to become the base function (such that spline part is zero, because weight = base + spline), `lamb_coefdiff` encourages the spline part to be smooth. 

When the distribution is skewed, you may want to use the `grid_eps` parameter, which specifies how adaptive of grid to sample distribution (by default grid_eps = 1 means uniform grid, grid_eps=0 means adaptive grid). You may want to set `KAN(..., grid_eps=0.01)` to use adaptive grid.

**Comment by KindXiaoming** [comment]
Hi, the real world distribution does not look too bad. The loss being 7.70e-01 looks suspicious. From the plot, looks like variable1 is totally ignored. Please try tune down `lamb`.

**Comment by kaneyxx** [comment]
Hi, report some experimental results I tried for discussion.
1. decreased `lamb`, I used 
```
model = KAN(width=[2,5,1], grid=5, k=3, seed=0, grid_eps=0.01)
model.train(dataset_m, opt="LBFGS", steps=steps, lamb=0.00001, lamb_l1=1., lamb_entropy=10., batch=1000); 
```
This looks helpful a bit. I can repeat prune/train cycle twice and obtain a [2,1,1] structure. However, it still cannot fit data well. The loss still stuck around 7.70e-1. 

2. increase `grid`, I used
```
for i in range(len(grids)):
    model = KAN(width=[2,5,1], grid=grids[i], k=3, seed=0, grid_eps=0.01,
                ).initialize_from_another_model(model.float(), dataset_m['train_input'].float())
    results = model.train(dataset_m, 
                        opt="LBFGS", 
                        steps=steps, 
                        stop_grid_update_step=30,
                        lamb=0.0001, 
                        lamb_l1=1., 
                        lamb_entropy=10.,
                        lr=0.1,
                        batch=1000)
```
to iterate [10, 20, 50] grids. The loss cannot decrease (it's really weird :/ ).

I'll keep trying and report. Btw, is Adam optimizer good for KAN?

**Comment by KindXiaoming** [comment]
please try changing `width=[2,5,1]` to `width=[2,1]`, which should work for your linear data.

**Comment by kaneyxx** [comment]
Yeah, I've also tried that structure. Here are some interesting findings.
1. Modify to `width=[2,1]`, loss 7.69e-1
![output_2_1_LBFGS_769e-1](https://github.com/KindXiaoming/pykan/assets/7875849/88442746-9802-42ea-a6cb-28a900c39b4e)
2. Modify to `width=[2,1]` and `opt="Adam"`, loss 7.69e-1
![output_2_1_adam_769e-1](https://github.com/KindXiaoming/pykan/assets/7875849/6875ba16-96b6-4df3-83ca-72cce80ef558)
3. I also checked the noisy level in real data. I found the real data is too noisy. Then, I filter noisy data and leave those clean data. (Absolute difference between inputs and target output is 0) 
I tried to fit the filtered data with method 1 and 2 above. Both of them still stuck around 7.3e-1.

**Comment by KindXiaoming** [comment]
That's really funny... what about set `lamb=0`?

**Comment by kaneyxx** [comment]
Thanks for this advice, `lamb=0` seems workable !? If I don't get it wrong, `lamb` is the overall penalty strength for whole regularization. That means we relax whole constraint for this model, right?
I've worked on another toy dataset with skewed distribution (I modified it from your dataset_generate function). Please take a look at this [notebook](https://colab.research.google.com/drive/19BrWfT3GlmPtT30cN8XlpNyjdfh_ijhP?usp=sharing). 
In this experiment, I changed `grid=3` `k=2`, `lamb=0` . The result looks on the good way but not fit well. Any further advice? I'll transfer to  real data later after fitting this toy set.

**Comment by KindXiaoming** [comment]
Yeah the outliers could be problematic for B-spline-based KANs. Could try 
* small grid_eps = 0.01
* small k = 1
* large grid (this is opposite to what I said earlier)
* lamb = 0
Or a combination of those

**Comment by kaneyxx** [comment]
One almost success case [link](https://colab.research.google.com/drive/1WGILk2e2NBnCMaPAljIkUW3rJinkqsyQ?usp=sharing)

I've tried these combinations, all of them work!
- small grid `grid=1` `grid=2` + small k `k=1`
- small grid `grid=1` + large k `k=3`
- large grid `grid=50` + small k `k=1`
- large grid `grid=50` + large k `k=3`

Only thing we need to do is to set `lamb=0` and `update_grid=False`
Actually, I have no idea why this work.
By the way, the result still a bit wrong $-1.154*\log{(4.6201x_1)}-0.203*\log{(11.2487x_2)+0.0001}+7.4222$ 
to the ground truth $-1.154*\log{x_1} - 0.203 * \log{x_2} + 5.165$

Another question about `fit_params()` function: 
When I use `model.fix_symbolic()` and pass the argument `fit_params_bool=True`, it will compute the coefficients of (a,b,c,d)
I found that this fitting function is sometimes incorrect. E.g.,
```
x = torch.linspace(-1,1,steps=100)
y = 1.0*(1.0*x + 0.0)**2 + 0.0
fit_params(x, y, lambda x: x**2)
>>> r2 is 1.0000004768371582
>>> (tensor([-8.3022e+00, -3.2000e-04,  1.4508e-02,  5.9605e-08]), tensor(1.0000))
```
I don't know if this function is causing my incorrect results.
Please give me some advices. Looking forward to further discussions.

**Comment by KindXiaoming** [comment]
hi, regarding the squared function, there are infinite combinations of (a,b,c,d) which can fit the data, e.g., x^2 = 1/a^2*(ax)^2, basically implying that (a,b,c,d) = (1,0,1,0) and (a,b,c,d) = (1/a^2,0,a,0) are equivalent.

**Comment by kaneyxx** [comment]
Hi, yeah, that makes sense! However, I am struggling with real data. I still have no idea why it doesn't work, even though the data is almost the same as the synthetic one. (The formula is totally the same. )

Recently, I have come up with some new questions:
1. Is there any chance we can integrate KANLayer with another model and still maintain its interpretability? For example, using several CNN layers to extract features and then a KAN layer for feature aggregation and interpretation through end-to-end training.
2. Sometimes the features are high-dimensional tensors. Do you have any suggestions on how to make them reasonable inputs for KAN?

**Comment by KindXiaoming** [comment]
My current hunch is that you need a 'converter' that converts any modality (vision, language etc.) to a "physical space" and then you can use KANs in the physical space/latent space (which might be low-dimensional so efficiency is not a problem). But I don't know what a 'converter' would look like - it might be as simple as any off-the-shelf encoder, but the interpretability of the encoder could be tricky; or we need to design completely new converters.

**Comment by kaneyxx** [comment]
Thanks for sharing your thoughts. This thread has helped me a lot. I'd like to explore more possibilities with KAN. I will close this issue as completed. If I find anything interesting or need further discussion, I will open a new issue or email you if it is not a bother. Hope we KAN do that :)
